init.lua

```diff
@@ -1,71 +1,82 @@
-		sneak={false},
-		aux1={false},
-		down={false},
-		up={false},
-		zoom={false},
-		dig={false},
-		place={false}
-	}
+--set up key store on join
+minetest.register_on_joinplayer(function(player, _)
+    local pname = player:get_player_name()
+    local controls_names = player:get_player_control()
+
+    --note: could hardcode this, but this is more future proof in case minetest adds more controls
+    controls.players[pname] = {}
+    for key, _ in pairs(controls_names) do
+        --[[
+            in theory the control value is false when they join, but hard coding just in case
+            consider changing this to named key table instead of numeric for better readability???
+        ]]
+        controls.players[pname][key] = {false}
+    end
 end)
 
-minetest.register_on_leaveplayer(function(player)
-	local name = player:get_player_name()
-	controls.players[name] = nil
+--discard when leaving
+minetest.register_on_leaveplayer(function(player, _)
+    local pname = player:get_player_name()
+    controls.players[pname] = nil
 end)
 
-minetest.register_globalstep(function()
-	for _, player in pairs(minetest.get_connected_players()) do
-		local player_name = player:get_player_name()
-		local player_controls = player:get_player_control()
-		for cname, cbool in pairs(player_controls) do
-			if not controls.players[player_name] then
-				-- player timed out but is still provided by get_connected_players(), disregard
-				break
-			end
+--event loop
+minetest.register_globalstep(function(dtime)
+    for _, player in pairs(minetest.get_connected_players()) do
+        local pname = player:get_player_name()
+        local pcontrols = player:get_player_control()
 
-			--Press a key
-			if cbool==true and controls.players[player_name][cname][1]==false then
-				for _, func in pairs(controls.registered_on_press) do
-					func(player, cname)
-				end
-				controls.players[player_name][cname] = {true, os.clock()}
-			elseif cbool==true and controls.players[player_name][cname][1]==true then
-				for _, func in pairs(controls.registered_on_hold) do
-					func(player, cname, os.clock()-controls.players[player_name][cname][2])
-				end
-			--Release a key
-			elseif cbool==false and controls.players[player_name][cname][1]==true then
-				for _, func in pairs(controls.registered_on_release) do
-					func(player, cname, os.clock()-controls.players[player_name][cname][2])
-				end
-				controls.players[player_name][cname] = {false}
-			end
-		end
-	end
+        if not controls.players[pname] then break end --safety check
+
+        --consider using minetest.get_us_time() instead of os.clock()? would need to convert to seconds however
+        for key, key_status in pairs(pcontrols) do
+            if key_status and not controls.players[pname][key][1] then
+                for _, callback in pairs(controls.registered_on_press) do
+                    callback(player, key)
+                end
+                controls.players[pname][key] = {true, os.clock()}
```

Comment:
I don't think `os.clock()` measures wall time, if I do `//lua local t = os.clock(); core.after(5, function() print(os.clock() - t) end)` I get a number somewhere around 1 (and not 5).

I wonder whether it would be better to do `controls.players[pname][key] = {true, 0}` here and add dtime to it before the below loop.

Reply:
see https://github.com/mt-mods/controls/blob/74401360ceee8ee617ea89aed05dd6cbc8d5a5b4/init.lua#L58 but that isnt guaranteed wall time either. cant dtime be funky if the server is under load/laggy iirc?